src/linear_algebra/matrix/spectrum.lean

```diff
@@ -0,0 +1,87 @@
+/-
+Copyright (c) 2022 Alexander Bentkamp. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Alexander Bentkamp
+-/
+import analysis.inner_product_space.spectrum
+import linear_algebra.matrix.hermitian
+
+/-! # Spectral theory of hermitian matrices
+
+This file proves the spectral theorem for matrices. The proof of the spectral theorem is based on
+the spectral theorem for linear maps (`diagonalization_basis_apply_self_apply`).
+
+## Tags
+
+spectral theorem, diagonalization theorem
+
+-/
+
+namespace matrix
+
+variables {𝕜 : Type*} [is_R_or_C 𝕜] [decidable_eq 𝕜] {n : Type*} [fintype n] [decidable_eq n]
+variables {A : matrix n n 𝕜}
+
+open_locale matrix
+
+namespace is_hermitian
+
+variables (hA : A.is_hermitian)
+
+/-- The eigenvalues of a hermitian matrix, indexed by `fin (fintype.card n)` where `n` is the index
+type of the matrix. -/
+noncomputable def eigenvalues₀ : fin (fintype.card n) → ℝ :=
+@inner_product_space.is_self_adjoint.eigenvalues 𝕜 _ _ (pi_Lp 2 (λ (_ : n), 𝕜)) _ A.to_lin'
+  (is_hermitian_iff_is_self_adjoint.1 hA) _ (fintype.card n) finrank_euclidean_space
+
+/-- The eigenvalues of a hermitian matrix, reusing the index `n` of the matrix entries. -/
+noncomputable def eigenvalues : n → ℝ :=
+  λ i, hA.eigenvalues₀ $ (fintype.equiv_of_card_eq (fintype.card_fin _)).symm i
+
+/-- A choice of an orthonormal basis of eigenvectors of a hermitian matrix. -/
+noncomputable def eigenvector_basis : basis n 𝕜 (n → 𝕜) :=
+  (@inner_product_space.is_self_adjoint.eigenvector_basis 𝕜 _ _
+  (pi_Lp 2 (λ (_ : n), 𝕜)) _ A.to_lin' (is_hermitian_iff_is_self_adjoint.1 hA) _ (fintype.card n)
+  finrank_euclidean_space).reindex (fintype.equiv_of_card_eq (fintype.card_fin _))
```

Comment:
```suggestion
(@inner_product_space.is_self_adjoint.eigenvector_basis 𝕜 _ _
    (pi_Lp 2 (λ (_ : n), 𝕜)) _ A.to_lin' (is_hermitian_iff_is_self_adjoint.1 hA) _
    (fintype.card n) finrank_euclidean_space).reindex
  (fintype.equiv_of_card_eq (fintype.card_fin _))
```

Reply:
All these `fintype.equiv_of_card_eq` suggest that it's worth generalizing the definitions about eigenvectors/-values in `inner_product_space` from `fin (finrank k E)` to `ι`, where `[fintype ι] (h : basis ι k E)`. And then define a new `sorted_eigenvalues` that lists them in increasing order as a `fin (finrank k E) → ℝ` tuple (as the TODO notes).

What do you think, would it be worth trying to do this change in a follow-up PR?